### PR TITLE
overlord: support for --quota-group option on snap install

### PR DESCRIPTION
**Pull request by Meulengracht**
This PR adds support for a new install option '--quota-group' which was the remaining feature of the journal quota spec. The challenge here was that I can't refer to the servicestate package from the snapstate package as that would unfortunately cause a cyclic dependency. Thus to solve this I've introduced a new task I can invoke anonomously from snapstate which does not require me to refer to types specified in servicestate.

### CODE REVIEW

File: overlord/servicestate/quota_handlers.go [hunk 1]
```diff
@@ -222,6 +220,80 @@ func addRestartServicesTasks(st *state.State, queueTask func(task *state.Task),
...
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	var quotaGroupName string
+	var snapNames []string
+	if err := t.Get("quota-on-install-snapnames", &snapNames); err != nil {
```
Comment (pedronis) [inline overlord/servicestate/quota_handlers.go:233]: we never put the name of the task in the parameter name, also I'm not sure why this should support multiple snaps? we have separate lanes per snaps in our install operation, so the normal thing would be to have one task per snap.

This then would be a single "snap-name"

File: overlord/servicestate/servicemgr.go [hunk 2]
```diff
@@ -62,6 +62,13 @@ func Manager(st *state.State, runner *state.TaskRunner) *ServiceManager {
 	// TODO: undo handler
 	runner.AddHandler("quota-control", m.doQuotaControl, nil)
 
+	// A separate, purpose specific task to handle support for adding
+	// a snap to a quota group upon install. We can't directly refer to
+	// the servicestate package, so this task takes care of calling quotaUpdate
+	// with the correct setup. This task also supports proper handling of
+	// failure during install and correctly removes the snap again.
+	runner.AddHandler("quota-on-install", m.doAddSnapToQuota, m.undoAddSnapToQuota)
```
Comment (pedronis) [inline overlord/servicestate/servicemgr.go:70]: is strange to be specifc about install here, as the task could be reused differently late.  Maybe if we want to make it clear that this is about quotas first:

"quota-add-snap",  doQuotaAddSnap", undoQuotaAddSnap

**Review by pedronis — COMMENTED** [review]
did a first quick pass on servicestate, some initial naming comments

File: overlord/servicestate/servicemgr.go [hunk 3]
```diff
@@ -62,6 +62,13 @@ func Manager(st *state.State, runner *state.TaskRunner) *ServiceManager {
...
+
 	snapstate.AddAffectedSnapsByKind("quota-control", quotaControlAffectedSnaps)
```
Comment (pedronis) [inline overlord/servicestate/servicemgr.go:72]: quota-add-snap will also need support for conflict detection

File: overlord/servicestate/servicemgr.go (hunk 2 above)
Comment (Meulengracht) [reply overlord/servicestate/servicemgr.go:70]: Ah ok, that makes sense, I had a lot trouble coming up with namings for the task and it's parameters, so I really appreciate the input here.

File: overlord/servicestate/quota_handlers.go [hunk 4]
```diff
@@ -222,6 +220,80 @@ func addRestartServicesTasks(st *state.State, queueTask func(task *state.Task),
...
+		return fmt.Errorf("internal error: cannot get quota-control-actions: %v", err)
+	}
+	if err := t.Get("quota-on-install-quotaname", &quotaGroupName); err != nil {
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply overlord/servicestate/quota_handlers.go:236]: Thank you for the input on naming, it's always really appreciated!

File: overlord/snapstate/snapmgr.go [hunk 5]
```diff
@@ -133,6 +133,10 @@ type SnapSetup struct {
 
 	// DisabledExposedHome is set if ~/Snap should not be used as $HOME.
 	DisableExposedHome bool `json:"disable-exposed-home,omitempty"`
+
+	// QuotaGroupName is set if the snap should be put into a quota group
+	// upon install
+	QuotaGroupName string `json:"quota-group-name,omitempty"`
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:139]: I think we can leave out the "-name" bit on the json side for consistency with the other json.

File: overlord/snapstate/flags.go [hunk 6]
```diff
@@ -98,6 +98,9 @@ type Flags struct {
 	// "per-snap" in case each snap is treated in a different
 	// transaction.
 	Transaction client.TransactionType `json:"transaction,omitempty"`
+
+	// QuotaGroupName
+	QuotaGroupName string `json:"quota-group-name,omitempty"`
```
Comment (pedronis) [inline overlord/snapstate/flags.go:103]: same

File: overlord/snapstate/snapstate.go [hunk 7]
```diff
@@ -531,6 +531,20 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		prev = installHook
 	}
 
+	if snapsup.QuotaGroupName != "" {
+		// This could result in doing 'setup-profiles' twice, but
+		// unfortunately we can't execute this code earlier as the snap
+		// needs to appear as installed first.
+		// XXX: Could we maybe move 'setup-profiles' to after this somehow? Or
+		// maybe just accept this could incur 'setup-profiles' twice?
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:539]: no, we can't move the whole setup-profiles, as it needs to happen before link-snap. setup-profiles uses Ensure* helpers quite a bit so disk writing is not a big issue, but is still a bit annoying. We need to look into this later

File: overlord/snapstate/snapstate.go [hunk 8]
```diff
@@ -531,6 +531,20 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
...
+		quotaControlTask := st.NewTask("quota-add-snap", fmt.Sprintf(i18n.G("Adding snap %q%s to quota group %q"),
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:540]: looking at the other NewTask:  s/Adding/Add

File: overlord/snapstate/snapstate_install_test.go [hunk 9]
```diff
@@ -4455,6 +4455,77 @@ func (s *snapmgrTestSuite) TestInstallPrerequisiteWithSameDeviceContext(c *C) {
...
 }
 
+func (s *snapmgrTestSuite) TestInstallQuotaGroup(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	var quotaWasCalled bool
+	s.o.TaskRunner().AddHandler("quota-add-snap", func(t *state.Task, _ *tomb.Tomb) error {
+		quotaWasCalled = true
+		return nil
```
Comment (pedronis) [inline overlord/snapstate/snapstate_install_test.go:4465]: this needs also to check that SnapSetup has the right value

File: overlord/snapstate/snapstate.go [hunk 10]
```diff
@@ -1044,6 +1058,7 @@ func InstallPath(st *state.State, si *snap.SideInfo, path, instanceName, channel
 		Type:               info.Type(),
 		PlugsOnly:          len(info.Slots) == 0,
 		InstanceKey:        info.InstanceKey,
+		QuotaGroupName:     flags.QuotaGroupName,
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1061]: this bit doesn't seem tested

File: daemon/api_snaps.go [hunk 11]
```diff
@@ -236,6 +237,7 @@ func (inst *snapInstruction) installFlags() (snapstate.Flags, error) {
 	if inst.IgnoreValidation {
 		flags.IgnoreValidation = true
 	}
+	flags.QuotaGroupName = inst.QuotaGroupName
```
Comment (pedronis) [inline daemon/api_snaps.go:240]: this needs some testing

File: cmd/snap/cmd_snap_op.go [hunk 12]
```diff
@@ -637,6 +638,7 @@ func (x *cmdInstall) Execute([]string) error {
 		IgnoreValidation: x.IgnoreValidation,
 		IgnoreRunning:    x.IgnoreRunning,
 		Transaction:      x.Transaction,
+		QuotaGroupName:   x.QuotaGroupName,
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go:641]: also this

**Review by pedronis — COMMENTED** [review]
did a pass for the places outside of servicestate, I need to re-review servicestate itself

File: overlord/servicestate/quota_handlers.go [hunk 13]
```diff
@@ -222,6 +220,86 @@ func addRestartServicesTasks(st *state.State, queueTask func(task *state.Task),
...
+	if err := t.Get("quota-name", &quotaName); err != nil {
+		return fmt.Errorf("internal error: cannot get quota-name: %v", err)
+	}
+
+	if err := CheckQuotaChangeConflictMany(st, []string{quotaName}); err != nil {
+		return err
+	}
+	if err := snapstate.CheckChangeConflictMany(st, []string{snapName}, ""); err != nil {
+		return err
+	}
```
Comment (pedronis) [inline overlord/servicestate/quota_handlers.go:245]: conflict checks are not something we do in task handlers, is something we need to do when we create the change

File: overlord/snapstate/snapstate.go [hunk 14]
```diff
@@ -531,6 +531,20 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		prev = installHook
 	}
 
+	if snapsup.QuotaGroupName != "" {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:534]: either here or one level up we need to do quota conflict checks

**Review by pedronis — CHANGES_REQUESTED** [review]
thanks for the changes in servicestate, the conflict checks inside handlers are unexpected though, we generally check for conflict before creating the actual changes.... for the other separate quote ops you can see it is done in quota_control.go, not quota_handlers.go

File: daemon/api_snaps.go [hunk 15]
```diff
@@ -203,6 +203,7 @@ type snapInstruction struct {
 	Snaps                  []string               `json:"snaps"`
```
Comment (pedronis) [inline daemon/api_snaps.go:203]: also the sideload code  (see corresponding file) supproting `snap install foo.snap` should also support this, there we have a multpart form

File: overlord/snapstate/flags.go (hunk 6 above)
Comment (Meulengracht) [reply overlord/snapstate/flags.go]: Fixed here as well, and updated the doc string which was empty

File: overlord/snapstate/snapstate_install_test.go (hunk 9 above)
Comment (Meulengracht) [reply overlord/snapstate/snapstate_install_test.go:4468]: Fixed!

File: daemon/api_snaps.go (hunk 11 above)
Comment (Meulengracht) [reply daemon/api_snaps.go:240]: Added unit tests for this!

File: cmd/snap/cmd_snap_op.go (hunk 12 above)
Comment (Meulengracht) [reply cmd/snap/cmd_snap_op.go:641]: Done!

**Review by mardy — DISMISSED** [review]
I'm not a fan of having most error messages starting with the `internal error: ` prefix, I'd rather remove it. But in any case, this LGTM!

File: overlord/snapstate/snapstate.go [hunk 16]
```diff
@@ -532,6 +532,24 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
...
+
+		// This could result in doing 'setup-profiles' twice, but
+		// unfortunately we can't execute this code earlier as the snap
+		// needs to appear as installed first.
+		// XXX: Could we maybe move 'setup-profiles' to after this somehow? Or
+		// maybe just accept this could incur 'setup-profiles' twice?
+		quotaControlTask := st.NewTask("quota-add-snap", fmt.Sprintf(i18n.G("Add snap %q%s to quota group %q"),
+			snapsup.InstanceName(), revisionStr, snapsup.QuotaGroupName))
+		quotaControlTask.Set("snap-name", snapsup.InstanceName())
+		quotaControlTask.Set("quota-name", snapsup.QuotaGroupName)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:548]: it feels that this code belong into a QuotaAddSnapTask helper set from servicestate into snapstate, has it happens with some of the hook stuff, see SetupInstallHook  etc for example

at that point the conflict stuff could move back there as well

File: daemon/api_snaps.go (hunk 15 above)
Comment (pedronis) [reply daemon/api_snaps.go:203]: this doesn't seem addressed yet

**Review by pedronis — COMMENTED** [review]
Thank you. A code placement remark, this has grown a bit and might need to grow more. I would split the overlord changes parts from the rest I think.

File: daemon/api_snaps.go (hunk 15 above)
Comment (Meulengracht) [reply daemon/api_snaps.go:203]: Yeah I missed this one, thanks for reminding me. Support for this is added in the latest commit

File: overlord/servicestate/quota_handlers.go (hunk 13 above)
Comment (Meulengracht) [reply overlord/servicestate/quota_handlers.go]: Fixed!

File: overlord/snapstate/snapstate.go (hunk 14 above)
Comment (Meulengracht) [reply overlord/snapstate/snapstate.go:535]: Fixed!

File: overlord/snapstate/snapstate.go (hunk 16 above)
Comment (Meulengracht) [reply overlord/snapstate/snapstate.go]: I did this as you mentioned, but I kept the conflict code seperated (but moved back to servicestate) and in similar style as the original conflict code, as I believe that makes more sense when we now have multiple places that need to check for conflicts.

File: overlord/servicestate/servicemgr.go [hunk 17]
```diff
@@ -61,9 +61,18 @@ func Manager(st *state.State, runner *state.TaskRunner) *ServiceManager {
 
 	// TODO: undo handler
 	runner.AddHandler("quota-control", m.doQuotaControl, nil)
-
+	AddAffectedQuotasByKind("quota-control", quotaControlAffectedQuotas)
 	snapstate.AddAffectedSnapsByKind("quota-control", quotaControlAffectedSnaps)
 
+	// A separate, purpose specific task to handle support for adding
+	// a snap to a quota group upon install. We can't directly refer to
+	// the servicestate package, so this task takes care of calling quotaUpdate
```
Comment (mardy) [inline overlord/servicestate/servicemgr.go:69]: It's unclear what this comment means, because we already are inside the `servicestate` package, so there should be no issue in using any of its members.

**Review by mardy — COMMENTED** [review]
No complaints on the code, but since it's rather hard to follow, I'm wondering if comments and renaming of methods can make it more clear.

File: overlord/servicestate/conflict.go [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+// An AffectedQuotasFunc returns a list of affected quota group names for the given supported task.
+type AffectedQuotasFunc func(*state.Task) ([]string, error)
+
+var (
+	affectedQuotasByKind = make(map[string]AffectedQuotasFunc)
+)
+
+// AddAffectedQuotasByKind registers an AffectedQuotasFunc for returning the affected quotas
+// for tasks of the given kind, to use in conflicts detection.
+func AddAffectedQuotasByKind(kind string, f AffectedQuotasFunc) {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/servicestate/conflict.go:37]: we have already functions that follow this pattern, so we want to change this it needs to happen separately. I go naming is very inconsistent about having Func in names of things that deal with functions.

File: overlord/servicestate/quota_handlers.go [hunk 19]
```diff
@@ -69,6 +69,10 @@ type QuotaControlAction struct {
 	ParentName string `json:"parent-name,omitempty"`
 }
 
+func init() {
+	snapstate.AddSnapToQuotaGroup = AddSnapToQuotaGroup
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/servicestate/quota_handlers.go:73]: this is another preexisting pattern, I don't want to diverge punctually on it here. We might revisit the pattern at a later point

**Comment by pedronis** [comment]
the overlord code has been split out to https://github.com/snapcore/snapd/pull/12066, that needs to be reviewed/landed first. Marking this as blocked for now

File: overlord/servicestate/quota_handlers.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/servicestate/quota_handlers.go:73]: to be clearer, I'm not against using Register instead of Add in the functions in the conflict mechanism, just that it should be its own change. It is also worth remembering that golang has types, so it's partly the case the purpose of a function can be understood from name + argument types and not only name.

[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/servicestate/quota_handlers.go:73]: (sorry I added a comment (now deleted) to the wrong comment thread)
